### PR TITLE
Implement caching of SdkEnvironments

### PR DESCRIPTION
**Pull request by jongerrish**
Replace rudimentary caching of SdkEnvironment in RobolectricTestRunner + SoftReferences (we don't want caching to depend on GC) with a basic LRU cache in EnvHolder. (We might want to rename EnvHolder to SdkEnvironmentCache maybe).

The cache is sized at the number of api levels supported by Robolectric since most test suites
will only ever have this number of environments. It is possible to have more environments than this
if multiple InstrumentingClassloaderConfigs are used which is rare, but is the case in the
Robolectric test suite. This is why we need to also include the test runner class name as part of the
key as a substitute for InstrumentingClassloaderConfig. We could consider refactoring InstrumentingClassloaderConfig to make it final, constructed using a builder and having properties (like instrumented packages / classes etc) that can be compared in a .equals() / .hashCode() sense. Alexander Blom also suggested only caching where config.getClass() == InstrumentingClassloaderConfig.class directly, so custom derived classes would not get cached. I'm open to ideas here.

Other cleanups included:-

* Don't pass the classlaoder into the ResourceExtractor in order to control the scope of references
to the classloader.
* Empty the map of loaded test classes at the end of the test as this keeps transient references to
the classloaders alive.

Move file -> manifest mapping from EnvHolder to test runner since EnvHolder is concerned with managing SdkConfig -> Classloader mappings.

### CODE REVIEW

**Comment by lexs** [comment]
I think this approach is fine, you still need some sort of caching even for custom InstrumentingClassloaderConfigs so you don't have to recreate the environment between test methods at least.

File: robolectric/src/main/java/org/robolectric/internal/EnvHolder.java [hunk 1]
```diff
@@ -1,22 +1,45 @@
...
+import org.robolectric.internal.dependency.DependencyResolver;
+import org.robolectric.util.Pair;
 
-import java.lang.ref.SoftReference;
-import java.util.HashMap;
+import java.net.URL;
+import java.util.LinkedHashMap;
 import java.util.Map;
 
 public class EnvHolder {
```
Comment (jongerrish) [inline robolectric/src/main/java/org/robolectric/internal/EnvHolder.java]: I'll probably rename this class in a follow up PR, perhaps InstrumentingClassloaderFactory ?

File: robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentingClassLoaderConfig.java [hunk 2]
```diff
@@ -35,48 +36,106 @@
 import java.util.ServiceLoader;
 import java.util.Set;
 
-import static java.util.Arrays.asList;
-
 /**
  * Configuration rules for {@link org.robolectric.internal.bytecode.InstrumentingClassLoader}.
  */
 public class InstrumentingClassLoaderConfig {
```
Comment (jongerrish) [inline robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentingClassLoaderConfig.java]: I was thinking to rename this InstrumentationConfig or InstrumentationConfigutation the name sounds a little more fluid.

Comment (erd) [inline robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentingClassLoaderConfig.java]: +1 for `InstrumentationConfiguration`

File: robolectric/src/main/java/org/robolectric/internal/EnvHolder.java (hunk 1 above)
Comment (erd) [inline robolectric/src/main/java/org/robolectric/internal/EnvHolder.java]: +1 for `InstrumentingClassLoaderFactory`

**Comment by lexs** [comment]
Note sure it's related to this but I'm getting a lot of classloader issues, like:
`java.lang.IllegalAccessError: class org.robolectric.res.ResBundle2Test$Values cannot access its superclass org.robolectric.res.ResBundle$Values`

**Comment by lexs** [comment]
Doesn't seem to be related to this work at all, nvm.

**Pull request merged**